bindings/rust/s2n-tls/src/renegotiate.rs

```diff
@@ -59,44 +155,224 @@ impl RenegotiateCallback for RenegotiateResponse {
+                // It is safe to return Pending here because `poll_recv` and
+                // `poll_send` are already responsible for clearing the input
+                // and output buffers respectively. The first one to succeed
+                // will block, but the second will wipe and begin renegotiation.
+                return Pending;
+            }
+            self.wipe_for_renegotiate()?;
+            self.renegotiate_state_mut().need_wipe = false;
+        }
+
         let mut blocked = s2n_blocked_status::NOT_BLOCKED;
-        let buf_len: isize = buf.len().try_into().unwrap_or(0);
-        let buf_ptr = buf.as_mut_ptr();
         let mut read: isize = 0;
-
-        let r = self.poll_negotiate_method(|conn| {
-            unsafe { s2n_renegotiate(conn.as_ptr(), buf_ptr, buf_len, &mut read, &mut blocked) }
-                .into_poll()
+        let result = self.poll_negotiate_method(|conn| {
+            unsafe {
+                s2n_renegotiate(
+                    conn.as_ptr(),
+                    buf_ptr as *mut u8,
+                    buf_len,
+                    &mut read,
+                    &mut blocked,
+                )
+            }
+            .into_poll()
         });
-        (r, read.try_into().unwrap())
+
+        if result.is_ready() {
+            self.renegotiate_state_mut().need_handshake = false
+        }
+        if read > 0 {
+            return Ready(Ok(read.try_into().unwrap()));
+        }
+        match result {
+            Ready(Ok(_)) => Pending,
+            Ready(Err(err)) => Ready(Err(err)),
+            Pending => Pending,
+        }
+    }
+
+    fn poll_renegotiate(&mut self, buf: &mut [u8]) -> Poll<Result<usize, Error>> {
+        let buf_len: isize = buf.len().try_into().map_err(|_| Error::INVALID_INPUT)?;
+        let buf_ptr = buf.as_mut_ptr() as *mut libc::c_void;
+        self.poll_renegotiate_raw(buf_ptr, buf_len)
+    }
+
+    /// Encrypts and sends data on a connection where
+    /// [negotiate](`Self::poll_negotiate`) has succeeded.
+    ///
+    /// Returns the number of bytes written, and may indicate a partial write.
+    ///
+    /// Automatically handles renegotiation.
+    pub fn poll_send(&mut self, buf: &[u8]) -> Poll<Result<usize, Error>> {
+        let mut blocked = s2n_blocked_status::NOT_BLOCKED;
+        let buf_len: isize = buf.len().try_into().map_err(|_| Error::INVALID_INPUT)?;
+        let buf_ptr = buf.as_ptr() as *const libc::c_void;
+
+        // If send is blocked, then we can't override poll_send to call
+        // poll_renegotiate until the application finishes retrying the send.
+        fn is_send_renegotiating(conn: &mut Connection) -> bool {
+            conn.is_renegotiating() && !conn.renegotiate_state().send_blocked
+        }
+
+        let is_renegotiating = is_send_renegotiating(self);
+        let result = if is_renegotiating {
+            let mut empty = [0; 0];
+            self.poll_renegotiate(&mut empty)
```

Comment:
Honestly, I think that's unavoidable. We can't change the poll_send interface.

The client can't send any application data after the ClientHello. However, the server can send ApplicationData after the ClientHello. So even if we verify that the ClientHello was definitely sent before blocking poll_send on renegotiation, the application could still get into a bad state because the receive call that triggered sending the ClientHello could be satisfied with application data received after the ClientHello is sent.

This situation isn't unique though. If you only call send, you'll miss the HelloRetry message altogether, never start renegotiation, and likely have the server just close the connection. Same with KeyUpdates. At least calling receive occasionally seems sane, and a client probably at least generally understands the call pattern with the server.

I've added some scary warnings.